src/Guzzle/SubjectRewritingMiddleware.php

```diff
@@ -0,0 +1,105 @@
+                }
+            }
+
+            $query = parse_query($request->getUri()->getQuery());
+            if (!empty($query['subject[]'])) {
+                $subjects = array_flip((array) $query['subject[]']);
+
+                foreach ($this->replacements as $replaced => $replacement) {
+                    if (isset($subjects[$replaced])) {
+                        unset($subjects[$replaced]);
+                        $subjects[$replacement['id']] = null;
+                    }
+                }
+
+                $uri = Uri::withoutQueryValue($uri, 'subject[]');
+                foreach (array_keys($subjects) as $subject) {
+                    $uri = Uri::withQueryValue($uri, 'subject[]', $subject);
+                }
+            }
+
+            $request = $request->withUri($uri);
+
+            /** @var PromiseInterface $promise */
+            $promise = $handler($request, $options);
+
+            return $promise->then(function (ResponseInterface $response) use ($request) {
+                $data = json_decode($response->getBody(), true);
+
+                if (0 === strncmp($request->getUri()->getPath(), '/subjects/', 10) || 0 === strncmp($request->getUri()->getPath(), '/highlights/', 12)) {
+                    // Nothing to do.
+                } elseif ($request->getUri()->getPath() === '/subjects') {
+                    $before = count($data['items']); // This won't work across pages.
+
+                    $data['items'] = array_filter($data['items'], function (array $subject) {
+                        return !in_array($subject['id'], array_keys($this->replacements));
+                    });
+
+                    $after = count($data['items']);
+
+                    $data['total'] = $data['total'] - ($before - $after);
+                } else {
+                    $visit = function ($item) use (&$visit) {
+                        if (!is_array($item)) {
+                            return $item;
+                        }
+
+                        if (!empty($item['id']) && !empty($item['name'])) {
+                            $keys = array_keys($item);
+                            sort($keys);
+                            if ($keys === ['id', 'name']) {
+                                if (isset($this->replacements[$item['id']])) {
+                                    return $this->replacements[$item['id']];
+                                }
+                            }
+                        }
+
+                        foreach ($item as $key => $value) {
+                            $item[$key] = $visit($value);
+
+                            if (in_array($key, ['expertises', 'subjects'], true)) {
+                                $item[$key] = array_unique($item[$key], SORT_REGULAR);
+                            }
+                        }
+
+                        return $item;
+                    };
+
+                    $data = $visit($data);
```

Comment:
Probably better to know where a subject could be and only investigate/change those.

Reply:
Also for performance. Subjects are used in many content types but I guess in only a few places in the schema, like a top-level field? https://github.com/elifesciences/api-raml/blob/e015430f48224ac294afc314f2d8416f36ff2c7a/dist/samples/blog-article-list/v1/first-page.json